### PR TITLE
Toponaming: support base feature in body correctly.

### LINKED ISSUE #14577: BaseFeature doesn't have TNP mitigations
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Do the following things:

1. Create two cubes
2. Shift the second cube so a quarter of it intersects the first
3. Fusion them
4. Create a PartDesign body
5. Drag the fusion in to PartDesign
6. Place a sketch on the second cube
7. Pad the sketch

It should look like this:

image

Now resize the second cube so instead of covering the quarter it instead intersects the cube. The pad breaks:

image

If instead of using a BaseFeature we use a ShapeBinder and repeat the process it works. I'm guessing this is because the attachment editor maps to the fusion itself rather than the binder?

### Full version info

```shell
OS: Arch Linux (i3/i3)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37668 (Git) AppImage
Build type: Release
Branch: main
Hash: ea68c5e88cdc80044ede482f6ad908820f4c1c82
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
Installed mods: 
  * Assembly4 0.50.4
  * Manipulator 1.5.0
  * fasteners 0.4.67
  * Help 1.0.3
```

### Subproject(s) affected?

None

### Anything else?

[basefeature.zip](https://github.com/user-attachments/files/15755525/basefeature.zip)
[shapebinder.zip](https://github.com/user-attachments/files/15755526/shapebinder.zip)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by bgbsww**
Builds on #14536 and fixes #14577.  Cleans up a couple of old tests and adds two new ones.

### CODE REVIEW

**Comment by chennes** [comment]
Nothing has been merged in between the time of the last successful CI run and the first failing CI run (as far as I can tell). @oursland or @adrianinsaval any ideas what's going on here? There's something failing that seems to be referencing a Qt6 definition, and none of our CI compiles should have anything involving Qt6 in them (yet!).

**Comment by adrianinsaval** [comment]
Vtk 9.3 on conda now pulls qt6, we can either pin it back to the previous version of vtk or see if we can move conda CI to use qt6. I'm not sure if the full dependency chain is ready yet. I'm gonna have a look tomorrow

**Comment by oursland** [comment]
VTK has been quite frustrating.  The `conda-forge` implementation of packaging of it puts a lot of unnecessary constraints, that could have been avoided if they had a multi-build of it with support for different dependencies.

Draft PR [#15000](https://github.com/FreeCAD/FreeCAD/pull/15000) pins to `vtk==9.2.6`.  If it succeeds in building, it should be un-drafted and merged.

File: src/Mod/PartDesign/App/Body.cpp [hunk 1]
```diff
@@ -190,6 +189,9 @@ bool Body::isSolidFeature(const App::DocumentObject *obj)
         return false;
     }
 
+    // if (obj == BaseFeature.getValue())
```
Comment (kadet1090) [inline src/Mod/PartDesign/App/Body.cpp:192]: Should this be left commented out?

File: src/Mod/Part/App/PartFeature.cpp [hunk 2]
```diff
@@ -483,13 +483,12 @@ static std::vector<std::pair<long, Data::MappedName>> getElementSource(App::Docu
...
-            //                static_cast<App::GeoFeature*>(owner)->getElementOwner(ret.back().second);
-            //                if (o)
-            //                    doc = o->getDocument();
-            //            }
+            if (owner->isDerivedFrom(App::GeoFeature::getClassTypeId())) {
+                auto o =
+                static_cast<App::GeoFeature*>(owner)->getElementOwner(ret.back().second);
+                if (o)
+                    doc = o->getDocument();
+            }
```
Comment (kadet1090) [inline src/Mod/Part/App/PartFeature.cpp:491]: This probably can be rewritten as (not sure how the formatting should look in that case):
```cpp
if (
    owner->isDerivedFrom(App::GeoFeature::getClassTypeId()) && 
    auto ownerGeoFeature = static_cast<App::GeoFeature*>(owner)->getElementOwner(ret.back().second)
) {
    doc = ownerGeoFeature->getDocument();
}
```

File: src/Mod/Part/App/PartFeature.cpp [hunk 3]
```diff
@@ -586,13 +585,12 @@ std::list<Data::HistoryItem> Feature::getElementHistory(App::DocumentObject* fea
...
-            //                static_cast<App::GeoFeature*>(feature)->getElementOwner(element);
-            //                if(owner)
-            //                    doc = owner->getDocument();
-            //            }
+            if(feature->isDerivedFrom(App::GeoFeature::getClassTypeId())) {
+                auto owner =
+                static_cast<App::GeoFeature*>(feature)->getElementOwner(element);
+                if(owner)
+                    doc = owner->getDocument();
+            }
```
Comment (kadet1090) [inline src/Mod/Part/App/PartFeature.cpp:593]: Same as previous comment

File: src/Mod/Part/App/PartFeature.cpp (hunk 2 above)
Comment (bgbsww) [reply src/Mod/Part/App/PartFeature.cpp:491]: You are right.  Thanks for looking at this with fresh eyes and leaving a review.

File: src/Mod/PartDesign/App/Body.cpp (hunk 1 above)
Comment (bgbsww) [reply src/Mod/PartDesign/App/Body.cpp:192]: Seems it's just an optimization that I brought over during code compare; makes sense to put it in.

Comment (bgbsww) [reply src/Mod/PartDesign/App/Body.cpp]: I take that back.  It doesn't work unless the method is converted from static, which triggers an avalanche of other issues.  That's why I originally removed the optimization, and now I've deleted it completely.

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 4]
```diff
@@ -285,14 +290,14 @@ void SketchObject::buildShape()
...
     // FIXME: Commented since ExternalGeometryFacade is not added
-    // for(i=2;i<ExternalGeo.getSize();++i) {
-    //     auto geo = ExternalGeo[i];
-    //     auto egf = ExternalGeometryFacade::getFacade(geo);
-    //     if(!egf->testFlag(ExternalGeometryExtension::Defining))
-    //         continue;
-    //     shapes.push_back(getEdge(geo, convertSubName(
-    //                     Data::IndexedName::fromConst("ExternalEdge", i-1), false).c_str()));
-    // }
+    for(i=2;i<ExternalGeo.size();++i) {
```
Comment (chennes) [inline src/Mod/Sketcher/App/SketchObject.cpp:293]: Can you clean up the sign comparison warning here?

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 5]
```diff
@@ -9420,6 +9425,115 @@ bool SketchObject::AutoLockTangencyAndPerpty(Constraint* cstr, bool bForce, bool
...
+App::DocumentObject *SketchObject::getSubObject(
+        const char *subname, PyObject **pyObj,
+        Base::Matrix4D *pmat, bool transform, int depth) const
+{
+    while(subname && *subname=='.') ++subname; // skip leading .
+    std::string sub;
+    const char *mapped = Data::isMappedElement(subname);
+    if(!subname || !subname[0])
+        return Part2DObject::getSubObject(subname,pyObj,pmat,transform,depth);
+    const char *element = Data::findElementName(subname);
```
Comment (chennes) [inline src/Mod/Sketcher/App/SketchObject.cpp:9437]: Unused "element"?

**Pull request merged**